### ISSUE TITLE
Custom cursors

### ISSUE THREAD

**Issue by mhemeryck** [psycopg/psycopg2#573]
I am currently working on a project where I need to pull data from one specific table where the layout can change over time. Over time, I have become quite comfortable with using ORMs such (SQLAlchemy, django-ORM), but as the layout can change over time, this is obviously not a solution.

My reasoning was thus to write my own custom row factory implementation, based on the custom cursors, such as the [`DictCursor`](http://initd.org/psycopg/docs/extras.html#psycopg2.extras.DictCursor) and the [`RealDictCursor`](http://initd.org/psycopg/docs/extras.html#psycopg2.extras.RealDictCursor).

By now, I have already succeeded in writing my own cursor / row factory class, where fetches result in objects being returned instead of `dict`s by directly subclassing the `RealDictCursor`. However, since there doesn't seem to be any documentation on writing my own cursor, I was wondering about some of the specifics:
* `__slots__ `: I suppose this is [an optimization](https://stackoverflow.com/a/28059785) required for lowering the memory footprint?
* `__getstate__` / `__setstate__`: this is for use in conjunction with the `__slots__` approach? Apparently, newer pickling protocols do support pickling with slotted instances though
* subclassing `DictCursorBase `: would you recommend always subclassing this for writing own cursors? Or directly from cursor? I suppose the magic here is to set / unset the `_prefetch`/ `_query_executed` flags and update the related instances accordingly, right?

**TL;DR**: what's the best approach to writing custom cursors?

**Comment by dvarrazzo** [comment]
Hello,

The only thing required is to subclass 'psycopg2.extensions.cursor'. You may subclass other classes if you find that handy, but the private methods and attributes, as well as Dictcursorbase, are not part of the public API so they are not documented and they may change without warnings: do it at your own risk (e.g. peg your program to a specific psycopg version, have tests).

The slots/getstate thing is an optimisation and not necessary for other subclasses to implement. 

IMO the best approach is to subclass 'cursor' and feel free to steal the bits of implementation you like from psycopg: copy/paste will likely give you less problems than subclassing.
